keps/sig-node/3675-resource-plugin-manager/README.md

```diff
@@ -0,0 +1,1328 @@
+dynamic and pluggable way of handling these resources.  
+
+Operating systems work by allowing drivers and pluggable resources.  This includes various
+policies in how cpu, memory, and devices are allocated.  Kubernetes can be viewed as being 
+an the operating system of the cloud.  Allowing specialty modules to address the use cases
+directly, rather than continuing to add complexity by continuing to modify the kubelet, 
+will allow the greatest scope of abilities while halting continued increases of complexity 
+within the core of the Kubelet.
+
+Users would like to be able to address the following use cases:
+
+* The ability to allow vendors to release vendor-specific managers for their hardware
+  and provide an interface while not having vendor-specific code within Kubelet.
+* Differentiate between types of cores and memory. <br>
+  Note - Dynamic resouce allocation does this with regular resources today.  We seek to
+  extend this ability.
+* Differentiate between different configurations of cores and memory, for instance cores
+  designated as performance versus those designated as efficiency cores 
+* Have custom plugins to optimize for particular types of workloads.  These plugins
+  may be built for performance, power reduction, cpu savings, et cetera.  <br>
+  Note:  Currently, there are very limited sets of available topology policies.  Every
+  new policy must be approved and be lockstep with the Kuberenetes release process.
+* Be able to hot-plug and test new resource managers.  
+* Be able to remove some of the complexity with current setup and, over time, reduce
+  the amount of code contained within Kubelet.  Instead, build a library with specific
+  needs.
+* Have a faster path to desired changes, without potentially impacting the core of 
+  Kubernetes with every policy change. <br>
+  Note that current solutions have been cropping up to allow for resource management
+  outside the core Kubelet. These solutions require turning off the Kubelet functionality 
+  and overriding current Kubelet allocation. We should provide a path otherwise. Many of these solutions are custom to particular companies. Many are not open source. The community should give a process to allow this functionality in core kubelet at a granular level, same as we have in many systems, such as HPC or telco.
+* Leverage additional parts from the pod spec information which could help make allocation decisions
+without the need to query the k8s controlplane.
+* Be able to do research, with minimum toil, on new policies and resource management strategies
+
+This design will also use gRPC, which is used for many other pluggable components within Kubernetes.  
+The hope is that it will also, as a side effect, allow a path to simplify the Kubelet as it is 
+today and move the complexity into external plugins.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+#### Alpha Goals
```

Comment:
> I've also expressed this before, but I will reiterate it again. I also strongly believe that we should not try to "bridge" the existing CPUManager with any CCI implemented CPUManagers as part of alpha. If a CPUManager policy of none is set, then requesting CPUs through a CCI implemented CPUManager is possible. If a CPUManager policy of static is set, then any pods that request CPUs through a CCI implemented CPUManager will fail. 

I really like this approach. The bridging behavior is a bit confusing for me and I haven't been able to understand how the handover would happen. Allowing CPU allocation with CCI driver when CPU manager policy is set to `none` means that we don't impact pods that rely on CPU Manager static policy for exclusive allocation of CPU and I would really like us to not impact existing capabilities 

> Moreover, we should only support running a single CCI implemented CPUManager (as a replacement for the builtin one) as part of alpha rather than building some internal component to reconcile allocations made by multiple ones.

I agree, let's keep things simple for alpha and provide a mechanism to introduce **a** pluggable solution for CPU management for custom use cases. State reconciliation across multiple such drivers is going to be very complicated and is best if we defer this to Beta.

Reply:
Kevin's approach outlined in the comment above looks fine to me as well.

Reply:
Just to summarize my understanding of the ask: the ask is if we can limit the scope of alpha version further with the objective : avoid bridging and api changes in alpha : possible approach exclusively choose between CCI Manager and CPU manager through a feature gate. If CCI feature is enabled and we have a CCI driver in the system handle them via the driver, otherwise pods should be able to start and run (equivalent to None policy) .

Reply:
We will capture that in the KEP, thank you

Reply:
> Just to summarize my understanding of the ask: the ask is if we can limit the scope of alpha version further with the objective : avoid bridging and api changes in alpha : possible approach exclusively choose between CCI Manager and CPU manager through a feature gate. If CCI feature is enabled and we have a CCI driver in the system handle them via the driver, otherwise pods should be able to start and run (equivalent to None policy) .

Yes, that is correct.

Reply:
captured.  Please review.